### PR TITLE
refactor: parquet store

### PR DESCRIPTION
**Pull request by domodwyer**
Refactors how code that produces/consumes Parquet files interacts with the object store.

I'll follow this up with another PR that decouples the parquet file codec (`RecordBatch(s) -> ParquetFile`) from the `ParquetStorage` and implement a streaming conversion, as well as tidying up the `ParquetStorage` API itself. This PR just lays the ground work.

Part of #4324. 

---

* refactor: tidy ParquetChunk constructor (5511be394)

      Removes two unused constructors for a ParquetChunk, and moves the bare fn
      constructor that is actually used to be an associated method (a conventional
      constructor).

* refactor: rename Storage -> ParquetStorage (deb29f7af)

      Renames the Storage type so the context is clear in usage (i.e. fn args),
      rather than having to rely on knowing the fully-qualified import path to know
      what the type stores.

* refactor: use ParquetStore instead of ObjectStore (678a1207c)

      Changes the code paths that interact with Parquet files in the object store to
      reference the ParquetStorage directly (DRY refactor).

      This change takes us from a dependency graph of:

                         ┌─────────────────┐
                         │                 │
                         ▼                 │
                 Parquet Consumer          │
                         │         ┌──────────────┐
                         ├────────▶│ParquetStorage│
                         ▼         └──────────────┘
                 ┌──────────────┐
                 │ ObjectStore  │
                 └──────────────┘
                         │
                    ┌────┴────┐
                    ▼         ▼
                  File       s3
                 System    (etc)

      to:

                     Parquet Consumer
                             │
                             ▼
                     ┌──────────────┐
                     │ParquetStorage│
                     └──────────────┘
                             │
                             ▼
                     ┌──────────────┐
                     │ ObjectStore  │
                     └──────────────┘
                             │
                        ┌────┴────┐
                        ▼         ▼
                      File       s3
                     System    (etc)

      With the ParquetStorage being solely responsible for managing interactions
      with the object store when dealing with Parquet files.

### CODE REVIEW

File: compactor/src/compact.rs [hunk 1]
```diff
@@ -804,7 +805,7 @@ impl Compactor {
             .expect("record_batches.is_empty was just checked")
             .schema();
 
-        let data = parquet_file::storage::Storage::new(Arc::clone(&object_store))
```
Comment (domodwyer) [inline compactor/src/compact.rs:807]: Prior to this PR, a `Storage` user would call `Storage::new(..)` passing an object store to encode the record batches into parquet bytes (not using the object store at all) and then...

File: compactor/src/compact.rs [hunk 2]
```diff
@@ -830,10 +830,9 @@ impl Compactor {
             metadata.partition_id,
             metadata.object_store_id,
         );
-        let path = path.object_store_path();
 
-        object_store
-            .put(&path, bytes)
```
Comment (domodwyer) [inline compactor/src/compact.rs:836]: ...use the object store it passed into `Storage::new(..)` to upload the parquet bytes itself.

File: compactor/src/compact.rs [hunk 3]
```diff
@@ -793,7 +794,7 @@ impl Compactor {
     async fn persist(
         metadata: &IoxMetadata,
         record_batches: Vec<RecordBatch>,
-        object_store: Arc<DynObjectStore>,
+        store: ParquetStorage,
```
Comment (domodwyer) [inline compactor/src/compact.rs:797]: Now a `ParquetStore` is initialised with the object store at the start of the program...

Comment (alamb) [reply compactor/src/compact.rs:797]: Sounds like a good design to me -- encapsulating storage details in `ParquetStorage` 👍

File: parquet_file/src/storage.rs [hunk 4]
```diff
@@ -249,17 +249,20 @@ impl Storage {
...
-                Self::download_and_scan_parquet(projection, path, object_store, tx.clone());
-
-            // If there was an error returned from download_and_scan_parquet send it back to the receiver.
-            if let Err(e) = download_result {
-                warn!(error=%e, "Parquet download & scan failed");
-                let e = ArrowError::ExternalError(Box::new(e));
-                if let Err(e) = tx.blocking_send(ArrowResult::Err(e)) {
-                    // if no one is listening, there is no one else to hear our screams
-                    debug!(%e, "Error sending result of download function. Receiver is closed.");
+        tokio::task::spawn_blocking({
```
Comment (alamb) [inline parquet_file/src/storage.rs:252]: I found the changes easier to see via https://github.com/influxdata/influxdb_iox/pull/4641/files?w=1

**Review by alamb — APPROVED** [review]
🧹  Nice and tidy

**Pull request merged**